### PR TITLE
Re-add old wallet restoration benchmark

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#1458 

# Overview

- [x] Restore old benchmark measuring the time it takes to restore wallets.
- [x] Make benchmark launch a cardano-node

# Comments
Successful run:
```
All results:
--
  | restore mainnet seq: 1016 s
  | restore mainnet 1% ownership: 3571 s
  | restore mainnet 2% ownership: 6687 s
```

https://buildkite.com/input-output-hk/cardano-wallet/builds/6536#e9097620-b59a-4b2d-9c44-c41982eea313

### CODE REVIEW

File: nix/haskell.nix [hunk 1]
```diff
@@ -85,6 +85,24 @@ let
...
 
+
+        # TODO: How do we bundle the config in the derivation, and
+        # do we even want to do this? Maybe there are better alternatives.
+        packages.cardano-wallet-byron.components.benchmarks.restore = {
+           build-tools = [ pkgs.makeWrapper ];
+           preBuild = ''
+             cat ${src + /lib/byron/test/data/bench-mainnet-configuration.yaml}
+             ls $out
+             cp ${src + /lib/byron/test/data/bench-mainnet-configuration.yaml} $out/config.yaml
```
Comment (Anviking) [inline nix/haskell.nix]: We — probably — need a custom config. 

1. To stop the node trying to write logs the the working directory where it doesn't have permissions
2. To disable the node's LiveView GUI

I suspect something like this could be done, but this in particular didn't work

Comment (Anviking) [reply nix/haskell.nix]: @rvl Does bundling the config like this sound like a sane idea to begin with, and do you know how to? (I suspect it's simple)

Comment (rvl) [reply nix/haskell.nix]: We could include our own logging config, but in this case I don't think we need to.

The liveview stuff probably won't affect the benchmark so it can stay. The other logging settings provided by cardano-node might help with debugging the benchmark.

The "trick" I used in cardano-launcher is to change working directory to somewhere where cardano-node is permitted to write its logs. So remove the `--run "cd $src"` line from below, because that forces the cwd to be a read-only location.

File: lib/byron/bench/Restore.hs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+
+    -- TODO: We might want to support testnet
+    let network = Mainnet
+
+    -- Reading setup from env vars.
+    --
+    -- TODO: Look over whether this is sane or not.
+    configs <- getEnv "BYRON_CONFIGS"
+    nodeDB <- getEnv "NODE_DB"
```
Comment (rvl) [inline lib/byron/bench/Restore.hs:184]: Maybe it would be easiest to set nodeDB to something like `./db` and socketPath to `./cardano-node.socket`. The logs will get dropped in `./logs` anyway. Then the script which executes the benchmark can take care of changing to the correct directory.

Comment (Anviking) [reply lib/byron/bench/Restore.hs:184]: I found it useful to be able to set NODE_DB when running locally, without nix.

`./cardano-node.socket` sounds sane though.

File: lib/byron/bench/Restore.hs [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+main = do
+    hSetBuffering stdout NoBuffering
+    hSetBuffering stderr NoBuffering
+
+    (_logCfg, tr) <- initBenchmarkLogging Info
+    installSignalHandlers (return ())
+
+    -- TODO: We might want to support testnet via a
+    -- `--benchmark-arguments "testnet"`
+    let network = Mainnet
```
Comment (Anviking) [inline lib/byron/bench/Restore.hs:172]: I think this can go into a separate PR.

**Comment by Anviking** [comment]
Btw, empty wallet, vs 1% vs 2% owned addresses using the logs from buildkite
Skärmavbild 2020-03-27 kl  13 15 59

File: lib/byron/bench/Restore.hs [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    -- it can be useful to set
+    --
+    --     export BYRON_CONFIGS=$CARDANO_NODE_DIR/configuration
+    --     export NODE_DB=$CARDANO_NODE_DIR/db
+    --
+    -- before running this benchmark manually with stack.
+    configs <- getEnv "BYRON_CONFIGS"
+    nodeDB <- getEnv "NODE_DB"
+
+    let genesisHash = "5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb"
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]: Note that this is already hard-coded in `Cardano.Wallet.Byron.Compatibility`, would be good limit the number of places where this gets written literally. 

```hs
let genesisHash = getGenesisBlockHash $ blockchainParameters @'Mainnet
```

File: lib/byron/bench/Restore.hs [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+    -- however.
+    let cmd = Command "cardano-node" args (return ()) NoStream NoStream
+    void $ withBackendProcess nullTracer cmd $ do
+        case network of
+            Testnet _ -> error "Testnet: not supported yet"
+            Mainnet -> do
+                prepareNode (Proxy @'Mainnet) socketPath
+                runBenchmarks
+                    [ bench ("restore " <> "mainnet" <> " seq")
+                        (bench_restoration @'Mainnet @ShelleyKey
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]: Doh! ShelleyKey aren't actually a thing on the Byron network. Byron network only uses `ByronKey` and `IcarusKey`! 
Benchmarks should be made for these two! The % one can be limited to only the IcarusKey though.

File: lib/byron/bench/Restore.hs [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+                        tl
+                        db
+                wallet <- unsafeRunExceptT $ W.createWallet w wid wname s
+                void $ forkIO $ unsafeRunExceptT $ W.restoreWallet @_ @s @t @k w wid
+                waitForWalletSync @n w wallet
+                (wallet', _, pending) <- unsafeRunExceptT $ W.readWallet w wid
+                sayErr "Wallet restored!"
+                sayErr . fmt $ "Balance: " +|| totalBalance pending wallet' ||+ " lovelace"
+                sayErr . fmt $
+                    "UTxO: " +|| Map.size (getUTxO $ totalUTxO pending wallet') ||+ " entries"
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]: What about printing the UTxO statistics also here? To get an overview of the distribution.

File: lib/byron/bench/Restore.hs [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+        slotDur = 20
+        epochDur = slotDur * 21600
+        ep = floor (d / epochDur)
+        idx = floor ((d - (fromIntegral ep) * epochDur) / slotDur)
+
+    startTime :: MonadFail m => m POSIXTime
+    startTime = pure
+        $ (\(StartTime t) -> utcTimeToPOSIXSeconds t)
+        $ getGenesisBlockDate
+        $ blockchainParameters @n
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]: I assume this function was taken from the old http-bridge implementation. We know have:

```
slotAt :: SlotParameters -> UTCTime -> Maybe SlotId
```

in `Cardano.Wallet.Primitive.Types` which can get us exactly the same result, what about using this instead of duplicating the logic :pray:  ?

File: lib/byron/bench/Restore.hs [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+
+logQuiet :: SlotId -> SlotId -> IO ()
+logQuiet _ _ = pure ()
+
+instance MaxSizeOf Address 'Mainnet ShelleyKey where
+    maxSizeOf = error "maxSizeOf should not be called by restore benchmark"
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]: It's more that "ShelleyKey" shouldn't be used in restoration benchmark :upside_down_face:

File: lib/byron/bench/Restore.hs [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+    -> FilePath
+       -- ^ Socket path
+    -> FilePath
+       -- ^ Log output
+    -> (WalletId, WalletName, s)
+    -> IO ()
+bench_restoration tracer socketPath progressLogFile (wid, wname, s) = do
+    let bp = blockchainParameters @n
+    let vData = versionData @n
+    let network = "mainnet" :: String
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs]:
```suggestion
    let network = networkDiscriminantVal @n
```

File: lib/byron/bench/Restore.hs (hunk 5 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs]: 👍 6632d448cfdc1fdedb7353cf7140aa77937d91f0

File: lib/byron/bench/Restore.hs (hunk 8 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs]: 6632d448cfdc1fdedb7353cf7140aa77937d91f0

File: lib/byron/bench/Restore.hs (hunk 9 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs]: 5dcfbfc4dcd3d375c9ada680eb7d69e3f809acda

File: lib/byron/bench/Restore.hs (hunk 7 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs]: ebf7978769b6876b7c31db914e41777780a058cc

File: lib/byron/bench/Restore.hs (hunk 4 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs]: 885ffee71e4e53fc0113ed19dcd2be872f242c59

File: lib/byron/bench/Restore.hs [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+
+{-------------------------------------------------------------------------------
+                                  Benchmarks
+-------------------------------------------------------------------------------}
+
+{-# ANN bench_restoration ("HLint: ignore Use camelCase" :: String) #-}
+bench_restoration
+    :: forall (n :: NetworkDiscriminant) (k :: Depth -> * -> *) s t.
+        ( IsOurs s Address
+        , IsOurs s ChimericAccount
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs:289]: Why is this constraint needed ?

File: lib/byron/bench/Restore.hs [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+            rngSeed = 0
+            s = mkRndState xprv rngSeed
+        in
+            (wid, wname, s)
+
+runBenchmarks :: [IO (Text, Double)] -> IO ()
+runBenchmarks bs = do
+    initializeTime
+    -- NOTE: Adding an artificial delay between successive runs to get a better
+    -- output for the heap profiling.
```
Comment (KtorZ) [inline lib/byron/bench/Restore.hs:254]: :+1:

**Review by KtorZ — APPROVED** [review]

**Comment by Anviking** [comment]
> What about printing the UTxO statistics also here?

Can go in a later PR.

ℹ️ Testing once more on normal CI 6fb36ec (should be dropped) before merging

**Comment by paweljakubas** [comment]
@Anviking  @KtorZ would be good also to add also latency benchmark in some future PR

**Comment by Anviking** [comment]
> would be good also to add also latency benchmark in some future PR

So a byron version of the existing (jormungandr) latency benchmark in the nightly. Sounds like a good idea.

File: lib/byron/bench/Restore.hs (hunk 10 above)
Comment (Anviking) [reply lib/byron/bench/Restore.hs:289]: For [`createWallet`](https://github.com/input-output-hk/cardano-wallet/blob/a3fecaccdb0600b0a5c76b1027441581cd2f43e6/lib/core/src/Cardano/Wallet.hs#L482), all the way to [`prefilterBlock`](https://github.com/input-output-hk/cardano-wallet/blob/a3fecaccdb0600b0a5c76b1027441581cd2f43e6/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L370-L377).
